<!DOCTYPE html>
<!-- START: inst/pkgdown/templates/layout.html --><!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><title>Introduction to Data Analysis with Python: Visualizing data with matplotlib and seaborn</title><meta name="viewport" content="width=device-width, initial-scale=1"><link rel="stylesheet" type="text/css" href="assets/styles.css"><script src="assets/scripts.js" type="text/javascript"></script><!-- mathjax --><script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      config: ["MMLorHTML.js"],
      jax: ["input/TeX","input/MathML","output/HTML-CSS","output/NativeMML", "output/PreviewHTML"],
      extensions: ["tex2jax.js","mml2jax.js","MathMenu.js","MathZoom.js", "fast-preview.js", "AssistiveMML.js", "a11y/accessibility-menu.js"],
      TeX: {
        extensions: ["AMSmath.js","AMSsymbols.js","noErrors.js","noUndefined.js"]
      },
      tex2jax: {
        inlineMath: [ ['$','$'], ['\\(', '\\)']],
        displayMath: [ ['$$','$$'], ['\\[', '\\]'] ],
        processEscapes: true
      }
    });
    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><!-- Responsive Favicon for The Carpentries --><link rel="apple-touch-icon" sizes="180x180" href="apple-touch-icon.png"><link rel="icon" type="image/png" sizes="32x32" href="favicon-32x32.png"><link rel="icon" type="image/png" sizes="16x16" href="favicon-16x16.png"><link rel="manifest" href="site.webmanifest"><link rel="mask-icon" href="safari-pinned-tab.svg" color="#5bbad5"><meta name="msapplication-TileColor" content="#da532c"><meta name="theme-color" content="#ffffff"></head><body>
    <header id="top" class="navbar navbar-expand-md navbar-light bg-white top-nav incubator"><a class="visually-hidden-focusable skip-link" href="#main-content">Skip to main content</a>
  <div class="container-fluid top-nav-container">
    <div class="col-md-6">
      <div class="large-logo">
        <img alt="Carpentries Incubator" src="assets/images/incubator-logo.svg"><abbr class="icon" title="This lesson is in the alpha phase, which means that it has been taught once and lesson authors are iterating on feedback." style="text-decoration: unset">
           
          <a href="https://cdh.carpentries.org/the-lesson-life-cycle.html#field-testing-alpha-stage" class="external-link alert-link" style="color: #383838">Alpha
            <i aria-hidden="true" class="icon" data-feather="alert-triangle" style="background: #FFC700; border-radius: 5px"></i>
          </a>
          <span class="visually-hidden">This lesson is in the alpha phase, which means that it has been taught once and lesson authors are iterating on feedback.</span>
        </abbr>
        
      </div>
    </div>
    <div class="selector-container ">
      
      
      <div class="dropdown">
        <button class="btn btn-secondary dropdown-toggle bordered-button" type="button" id="dropdownMenu1" data-bs-toggle="dropdown" aria-expanded="false">
          <i aria-hidden="true" class="icon" data-feather="eye"></i> Learner View <i data-feather="chevron-down"></i>
        </button>
        <ul class="dropdown-menu" aria-labelledby="dropdownMenu1"><li><button class="dropdown-item" type="button" onclick="window.location.href='instructor/13-plotting.html';">Instructor View</button></li>
        </ul></div>
    </div>
  </div>
  <hr></header><nav class="navbar navbar-expand-xl navbar-light bg-white bottom-nav incubator" aria-label="Main Navigation"><div class="container-fluid nav-container">
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
      <span class="menu-title">Menu</span>
    </button>
    <div class="nav-logo">
      <img class="small-logo" alt="Carpentries Incubator" src="assets/images/incubator-logo-sm.svg"></div>
    <div class="lesson-title-md">
      Introduction to Data Analysis with Python
    </div>
    <div class="search-icon-sm">
      <!-- TODO: do not show until we have search
      <i role="img" aria-label="search button" data-feather="search"></i>
      -->
    </div>
    <div class="desktop-nav">
      <ul class="navbar-nav me-auto mb-2 mb-lg-0"><li class="nav-item">
          <span class="lesson-title">
            Introduction to Data Analysis with Python
          </span>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="key-points.html">Key Points</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="reference.html#glossary">Glossary</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="profiles.html">Learner Profiles</a>
        </li>
        <li class="nav-item dropdown">
          <button class="nav-link dropdown-toggle" id="navbarDropdown" data-bs-toggle="dropdown" aria-expanded="false">
            More <i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu" aria-labelledby="navbarDropdown"><li><a class="dropdown-item" href="reference.html">Reference</a></li>
          </ul></li>
      </ul></div>
    <form class="d-flex col-md-2 search-form">
      <fieldset disabled><input class="form-control me-2 searchbox" type="search" placeholder="Search" aria-label="Search"><button class="btn btn-outline-success tablet-search-button" type="submit">
          <i class="search-icon" data-feather="search" role="img" aria-label="search button"></i>
        </button>
      </fieldset></form>
  </div><!--/div.container-fluid -->
</nav><div class="col-md-12 mobile-title">
  Introduction to Data Analysis with Python
</div>

<aside class="col-md-12 lesson-progress"><div style="width: 84%" class="percentage">
    84%
  </div>
  <div class="progress incubator">
    <div class="progress-bar incubator" role="progressbar" style="width: 84%" aria-valuenow="84" aria-label="Lesson Progress" aria-valuemin="0" aria-valuemax="100">
    </div>
  </div>
</aside><div class="container">
      <div class="row">
        <!-- START: inst/pkgdown/templates/navbar.html -->
  <div id="sidebar-col" class="col-lg-4">
    <div id="sidebar" class="sidebar">
      <nav aria-labelledby="flush-headingEleven"><button role="button" aria-label="close menu" alt="close menu" aria-expanded="true" aria-controls="sidebar" class="collapse-toggle"><i class="search-icon" data-feather="x" role="img"></i></button>
        <div class="sidebar-inner">
          <div class="row mobile-row">
            <div class="col">
              <div class="sidenav-view-selector">
                <div class="accordion accordion-flush" id="accordionFlush9">
                  <div class="accordion-item">
                    <h2 class="accordion-header" id="flush-headingNine">
                      <button class="accordion-button collapsed" id="instructor" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseNine" aria-expanded="false" aria-controls="flush-collapseNine">
                        <i id="eye" aria-hidden="true" class="icon" data-feather="eye"></i> Learner View
                      </button>
                    </h2>
                    <div id="flush-collapseNine" class="accordion-collapse collapse" aria-labelledby="flush-headingNine" data-bs-parent="#accordionFlush2">
                      <div class="accordion-body">
                        <a href="instructor/13-plotting.html">Instructor View</a>
                      </div>
                    </div>
                  </div><!--/div.accordion-item-->
                </div><!--/div.accordion-flush-->
              </div><!--div.sidenav-view-selector -->
            </div><!--/div.col -->
      
            <hr></div><!--/div.mobile-row -->

          <div class="accordion accordion-flush" id="accordionFlush11">
            <div class="accordion-item">

              <button id="chapters" class="accordion-button show" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseEleven" aria-expanded="false" aria-controls="flush-collapseEleven">
                <h2 class="accordion-header chapters" id="flush-headingEleven">
                  EPISODES
                </h2>
              </button>
              <div id="flush-collapseEleven" class="accordion-collapse show collapse" aria-labelledby="flush-headingEleven" data-bs-parent="#accordionFlush11">

                <div class="accordion-body">
                  <div class="accordion accordion-flush" id="accordionFlush1">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading1">
        <a href="index.html">Summary and Setup</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush2">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading2">
        <a href="01-introduction.html">1. Welcome to Python</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush3">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading3">
        <a href="02-variables.html">2. Variables in Python</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush4">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading4">
        <a href="03-types.html">3. Basic Types</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush5">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading5">
        <a href="04-using-functions.html">4. Built-in Functions and Help</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush6">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading6">
        <a href="home-01-strings.html">5. String Manipulation</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush7">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading7">
        <a href="home-02-objects.html">6. Using Objects</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush8">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading8">
        <a href="05-lists.html">7. Lists</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush9">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading9">
        <a href="06-loops.html">8. For Loops</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush10">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading10">
        <a href="07-packages.html">9. Libraries</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush11">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading11">
        <a href="08-loading.html">10. Reading tabular data</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush12">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading12">
        <a href="home-03-conda.html">11. Managing Python Environments</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush13">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading13">
        <a href="home-04-dictionaries.html">12. Dictionaries</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush14">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading14">
        <a href="09-conditionals.html">13. Conditionals</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush15">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading15">
        <a href="10-dataframes.html">14. Pandas DataFrames</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush16">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading16">
        <a href="11-writing-functions.html">15. Writing Functions</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush17">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading17">
        <a href="12-scipy.html">16. Perform Statistical Tests with Scipy</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush18">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading18">
        <a href="home-05-reshaping-data.html">17. Reshaping Data</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush19">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading19">
        <a href="home-06-combining-data.html">18. Combining Data</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlushcurrent">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-headingcurrent">
      <button class="accordion-button" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapsecurrent" aria-expanded="true" aria-controls="flush-collapsecurrent">
        <span class="visually-hidden">Current Chapter</span>
        <span class="current-chapter">
        19. Visualizing data with matplotlib and seaborn
        </span>
      </button>
    </div><!--/div.accordion-header-->
        
    <div id="flush-collapsecurrent" class="accordion-collapse collapse show" aria-labelledby="flush-headingcurrent" data-bs-parent="#accordionFlushcurrent">
      <div class="accordion-body">
        <ul><li><a href="#matplotlib-is-the-most-widely-used-scientific-plotting-library-in-python."></a><a href="https://matplotlib.org/" class="external-link"><code>matplotlib</code></a> is the
most widely used scientific plotting library in Python.</li>
<li><a href="#plot-data-directly-from-a-pandas-dataframe.">Plot data directly from a </a><a href="https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.html" class="external-link"><code>Pandas dataframe</code></a>.</li>
<li><a href="#select-and-transform-data-then-plot-it.">Select and transform data, then plot it.</a></li>
<li><a href="#many-plot-options-are-available.">Many plot options are available.</a></li>
<li><a href="#data-can-also-be-plotted-by-calling-the-matplotlib-plot-function-directly.">Data can also be plotted by calling the <code>matplotlib</code>
<code>plot</code> function directly.</a></li>
<li><a href="#you-can-plot-many-sets-of-data-together.">You can plot many sets of data together.</a></li>
<li><a href="#seaborn-provides-more-integrated-plotting-with-pandas-and-useful-defaults">Seaborn provides more integrated plotting with Pandas and useful
defaults</a></li>
<li><a href="#scenario-log-foldchange-scatterplot">Scenario: log foldchange scatterplot</a></li>
<li><a href="#heatmaps">Heatmaps</a></li>
        </ul></div><!--/div.accordion-body-->
    </div><!--/div.accordion-collapse-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush21">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading21">
        <a href="14-sklearn.html">20. Perform machine learning with Scikit-learn</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush22">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading22">
        <a href="15-biomart.html">21. ID mapping using mygene</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

                </div>
              </div>
            </div>

            <hr class="half-width"><div class="accordion accordion-flush resources" id="accordionFlush12">
              <div class="accordion-item">
                <h2 class="accordion-header" id="flush-headingTwelve">
                  <button class="accordion-button collapsed" id="resources" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseTwelve" aria-expanded="false" aria-controls="flush-collapseTwelve">
                    RESOURCES
                  </button>
                </h2>
                <div id="flush-collapseTwelve" class="accordion-collapse collapse" aria-labelledby="flush-headingTwelve" data-bs-parent="#accordionFlush12">
                  <div class="accordion-body">
                    <ul><li>
                        <a href="key-points.html">Key Points</a>
                      </li>
                      <li>
                        <a href="reference.html#glossary">Glossary</a>
                      </li>
                      <li>
                        <a href="profiles.html">Learner Profiles</a>
                      </li>
                      <li><a href="reference.html">Reference</a></li>
                    </ul></div>
                </div>
              </div>
            </div>
            <hr class="half-width resources"><a href="aio.html">See all in one page</a>

            <hr class="d-none d-sm-block d-md-none"><div class="d-grid gap-1">
            
            </div>
          </div><!-- /div.accordion -->
        </div><!-- /div.sidebar-inner -->
      </nav></div><!-- /div.sidebar -->
  </div><!-- /div.sidebar-col -->
<!-- END:   inst/pkgdown/templates/navbar.html-->

        <!-- START: inst/pkgdown/templates/content-instructor.html -->
  <div class="col-xl-8 col-lg-12 primary-content">
    <nav class="lesson-content mx-md-4" aria-label="Previous and Next Chapter"><!-- content for small screens --><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="home-06-combining-data.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="14-sklearn.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="home-06-combining-data.html" rel="prev"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous: Combining Data</a>
        <a class="chapter-link float-end" href="14-sklearn.html" rel="next">Next: Perform machine... <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <hr></nav><main id="main-content" class="main-content"><div class="container lesson-content">
        <h1>Visualizing data with matplotlib and seaborn</h1>
        <p> Last updated on 2023-04-27 | 
        
        <a href="https://github.com/ccb-hms/workbench-python-workshop/edit/main/episodes/13-plotting.md" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
        
        
        
        <div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right"> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>

        

<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul><li>How can I plot my data?</li>
<li>How can I save my plot for publishing?</li>
</ul></div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul><li>Visualize list-type data using matplotlib.</li>
<li>Visualize dataframe data with matplotlib and seaborn.</li>
<li>Customize plot aesthetics.</li>
<li>Create publication-ready plots for a particular context.</li>
</ul></div>
</div>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Keypoints<a class="anchor" aria-label="anchor" href="#keypoints1"></a>
</h3>
<div class="callout-content">
<ul><li>
<a href="https://matplotlib.org/" class="external-link"><code>matplotlib</code></a> is the
most widely used scientific plotting library in Python.</li>
<li>Plot data directly from a Pandas dataframe.</li>
<li>Select and transform data, then plot it.</li>
<li>Many styles of plot are available: see the <a href="https://python-graph-gallery.com/matplotlib/" class="external-link">Python Graph
Gallery</a> for more options.</li>
<li>Seaborn extends matplotlib and provides useful defaults and
integration with dataframes.</li>
</ul></div>
</div>
</div>
<section id="matplotlib-is-the-most-widely-used-scientific-plotting-library-in-python."><h2 class="section-heading">
<a href="https://matplotlib.org/" class="external-link"><code>matplotlib</code></a> is the
most widely used scientific plotting library in Python.<a class="anchor" aria-label="anchor" href="#matplotlib-is-the-most-widely-used-scientific-plotting-library-in-python."></a>
</h2>
<hr class="half-width"><ul><li>Commonly use a sub-library called <a href="https://matplotlib.org/stable/api/_as_gen/matplotlib.pyplot.html#module-matplotlib.pyplot" class="external-link"><code>matplotlib.pyplot</code></a>.</li>
<li>The Jupyter Notebook will render plots inline by default.</li>
</ul><div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">PYTHON<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode python" tabindex="0"><code class="sourceCode python"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> matplotlib.pyplot <span class="im">as</span> plt</span></code></pre>
</div>
<ul><li>Simple plots are then (fairly) simple to create.</li>
</ul><div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">PYTHON<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode python" tabindex="0"><code class="sourceCode python"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>time <span class="op">=</span> [<span class="dv">0</span>, <span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>]</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>position <span class="op">=</span> [<span class="dv">0</span>, <span class="dv">100</span>, <span class="dv">200</span>, <span class="dv">300</span>]</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>plt.plot(time, position)</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>plt.xlabel(<span class="st">'Time (hr)'</span>)</span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a>plt.ylabel(<span class="st">'Position (km)'</span>)</span></code></pre>
</div>
<figure><img src="fig/13_1_simpleplot.png" alt="" class="figure mx-auto d-block"><figcaption>Simple Position-Time Plot</figcaption></figure><div id="display-all-open-figures" class="callout callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="display-all-open-figures" class="callout-inner">
<h3 class="callout-title">Display All Open Figures<a class="anchor" aria-label="anchor" href="#display-all-open-figures"></a>
</h3>
<div class="callout-content">
<p>In our Jupyter Notebook example, running the cell should generate the
figure directly below the code. The figure is also included in the
Notebook document for future viewing. However, other Python environments
like an interactive Python session started from a terminal or a Python
script executed via the command line require an additional command to
display the figure.</p>
<p>Instruct <code>matplotlib</code> to show a figure:</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">PYTHON<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode python" tabindex="0"><code class="sourceCode python"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code></pre>
</div>
<p>This command can also be used within a Notebook - for instance, to
display multiple figures if several are created by a single cell.</p>
</div>
</div>
</div>
</section><section id="plot-data-directly-from-a-pandas-dataframe."><h2 class="section-heading">Plot data directly from a <a href="https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.html" class="external-link"><code>Pandas dataframe</code></a>.<a class="anchor" aria-label="anchor" href="#plot-data-directly-from-a-pandas-dataframe."></a>
</h2>
<hr class="half-width"><ul><li>We can also plot <a href="https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.html" class="external-link">Pandas
dataframes</a>.</li>
<li>This implicitly uses <a href="https://matplotlib.org/stable/api/_as_gen/matplotlib.pyplot.html#module-matplotlib.pyplot" class="external-link"><code>matplotlib.pyplot</code></a>.</li>
</ul><div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">PYTHON<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode python" tabindex="0"><code class="sourceCode python"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>url <span class="op">=</span> <span class="st">"https://raw.githubusercontent.com/ccb-hms/workbench-python-workshop/main/episodes/data/rnaseq.csv"</span></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>rnaseq_df <span class="op">=</span> pd.read_csv(url, index_col<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>rnaseq_df.loc[:,<span class="st">'expression'</span>].plot(kind <span class="op">=</span> <span class="st">'hist'</span>)</span></code></pre>
</div>
<figure><img src="fig/13_2_pandahist.png" alt="" class="figure mx-auto d-block"><figcaption>Expression histogram</figcaption></figure></section><section id="select-and-transform-data-then-plot-it."><h2 class="section-heading">Select and transform data, then plot it.<a class="anchor" aria-label="anchor" href="#select-and-transform-data-then-plot-it."></a>
</h2>
<hr class="half-width"><ul><li>By default, <a href="https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.plot.html#pandas.DataFrame.plot" class="external-link"><code>DataFrame.plot</code></a>
plots with the rows as the X axis.</li>
<li>We can transform the data to plot multiple samples.</li>
</ul><div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">PYTHON<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode python" tabindex="0"><code class="sourceCode python"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>expression_matrix <span class="op">=</span> rnaseq_df.pivot_table(columns <span class="op">=</span> <span class="st">"time"</span>, values <span class="op">=</span> <span class="st">"expression"</span>, index<span class="op">=</span><span class="st">"gene"</span>)</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>expression_matrix.plot(kind<span class="op">=</span><span class="st">"box"</span>, logy<span class="op">=</span><span class="va">True</span>)</span></code></pre>
</div>
<figure><img src="fig/13_3_timebox.png" alt="" class="figure mx-auto d-block"><figcaption>Boxplot by timepoint</figcaption></figure></section><section id="many-plot-options-are-available."><h2 class="section-heading">Many plot options are available.<a class="anchor" aria-label="anchor" href="#many-plot-options-are-available."></a>
</h2>
<hr class="half-width"><ul><li>For example, we can use the <code>subplots</code> argument to
separate our plot automatically by column.</li>
</ul><div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">PYTHON<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode python" tabindex="0"><code class="sourceCode python"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>expression_matrix.plot(kind<span class="op">=</span><span class="st">"hist"</span>, subplots<span class="op">=</span><span class="va">True</span>)</span></code></pre>
</div>
<figure><img src="fig/13_4_timehist_sub.png" alt="" class="figure mx-auto d-block"><figcaption>Histogram by timpoint as subplots.</figcaption></figure></section><section id="data-can-also-be-plotted-by-calling-the-matplotlib-plot-function-directly."><h2 class="section-heading">Data can also be plotted by calling the <code>matplotlib</code>
<code>plot</code> function directly.<a class="anchor" aria-label="anchor" href="#data-can-also-be-plotted-by-calling-the-matplotlib-plot-function-directly."></a>
</h2>
<hr class="half-width"><ul><li>The command is <code>plt.plot(x, y)</code>
</li>
<li>The color and format of markers can also be specified as an
additional optional argument e.g., <code>b-</code> is a blue line,
<code>g--</code> is a green dashed line.</li>
</ul><div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">PYTHON<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode python" tabindex="0"><code class="sourceCode python"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>expression_asl <span class="op">=</span> expression_matrix.loc[<span class="st">'Asl'</span>]</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>plt.plot(expression_asl, <span class="st">'g--'</span>)</span></code></pre>
</div>
<figure><img src="fig/13_5_asltime.png" alt="" class="figure mx-auto d-block"><figcaption>Asl expression over time</figcaption></figure></section><section id="you-can-plot-many-sets-of-data-together."><h2 class="section-heading">You can plot many sets of data together.<a class="anchor" aria-label="anchor" href="#you-can-plot-many-sets-of-data-together."></a>
</h2>
<hr class="half-width"><div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">PYTHON<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode python" tabindex="0"><code class="sourceCode python"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>plt.plot(expression_matrix.loc[<span class="st">'Asl'</span>], <span class="st">'g--'</span>, label <span class="op">=</span> <span class="st">'Asl'</span>)</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>plt.plot(expression_matrix.loc[<span class="st">'Apod'</span>], <span class="st">'r-'</span>, label <span class="op">=</span> <span class="st">'Apod'</span>)</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>plt.plot(expression_matrix.loc[<span class="st">'Cyp2d22'</span>], <span class="st">'b-'</span>, label <span class="op">=</span> <span class="st">'Cyp2d22'</span>)</span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>plt.plot(expression_matrix.loc[<span class="st">'Klk6'</span>], <span class="st">'b--'</span>, label <span class="op">=</span> <span class="st">'Klk6'</span>)</span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Create legend.</span></span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a>plt.legend(loc<span class="op">=</span><span class="st">'upper left'</span>)</span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a>plt.xlabel(<span class="st">'Time (days)'</span>)</span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a>plt.ylabel(<span class="st">'Mean Expression'</span>)</span></code></pre>
</div>
<figure><img src="fig/13_6_multiline.png" alt="" class="figure mx-auto d-block"><figcaption>Multiple lines on the same plot with a legend</figcaption></figure><div id="adding-a-legend" class="callout callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="adding-a-legend" class="callout-inner">
<h3 class="callout-title">Adding a Legend<a class="anchor" aria-label="anchor" href="#adding-a-legend"></a>
</h3>
<div class="callout-content">
<p>Often when plotting multiple datasets on the same figure it is
desirable to have a legend describing the data.</p>
<p>This can be done in <code>matplotlib</code> in two stages:</p>
<ul><li>Provide a label for each dataset in the figure:</li>
</ul><div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">PYTHON<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode python" tabindex="0"><code class="sourceCode python"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>plt.plot(expression_matrix.loc[<span class="st">'Asl'</span>], <span class="st">'g--'</span>, label <span class="op">=</span> <span class="st">'Asl'</span>)</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>plt.plot(expression_matrix.loc[<span class="st">'Apod'</span>], <span class="st">'r-'</span>, label <span class="op">=</span> <span class="st">'Apod'</span>)</span></code></pre>
</div>
<ul><li>Instruct <code>matplotlib</code> to create the legend.</li>
</ul><div class="codewrapper sourceCode" id="cb10">
<h3 class="code-label">PYTHON<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode python" tabindex="0"><code class="sourceCode python"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>plt.legend()</span></code></pre>
</div>
<p>By default matplotlib will attempt to place the legend in a suitable
position. If you would rather specify a position this can be done with
the <code>loc=</code> argument, e.g to place the legend in the upper
left corner of the plot, specify <code>loc='upper left'</code></p>
</div>
</div>
</div>
</section><section id="seaborn-provides-more-integrated-plotting-with-pandas-and-useful-defaults"><h2 class="section-heading">Seaborn provides more integrated plotting with Pandas and useful
defaults<a class="anchor" aria-label="anchor" href="#seaborn-provides-more-integrated-plotting-with-pandas-and-useful-defaults"></a>
</h2>
<hr class="half-width"><p><a href="https://seaborn.pydata.org/index.html" class="external-link">Seaborn</a> is a
plotting library build ontop of <code>matplotlib.pyplot</code>. It
provides higher-level functions for creating high quality data
visualizations, and easily integrates with pandas dataframes.</p>
<p>Let’s see what a boxplot looks like using seaborn.</p>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">PYTHON<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode python" tabindex="0"><code class="sourceCode python"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> seaborn <span class="im">as</span> sns</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> numpy <span class="im">as</span> np</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a><span class="co">#Use seaborn default styling</span></span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a>sns.set_theme()</span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a><span class="co">#Add a log expression column with a pseudocount of 1</span></span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true" tabindex="-1"></a>rnaseq_df <span class="op">=</span> rnaseq_df.assign(log_exp <span class="op">=</span> np.log2(rnaseq_df[<span class="st">"expression"</span>]<span class="op">+</span><span class="dv">1</span>))</span>
<span id="cb11-9"><a href="#cb11-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-10"><a href="#cb11-10" aria-hidden="true" tabindex="-1"></a>sns.boxplot(data <span class="op">=</span> rnaseq_df, x <span class="op">=</span> <span class="st">"time"</span>, y <span class="op">=</span> <span class="st">"log_exp"</span>)</span></code></pre>
</div>
<figure><img src="fig/13_7_simpbox.png" alt="" class="figure mx-auto d-block"><figcaption>A simple boxplot using seaborn</figcaption></figure><p>Instead of providing data to x and y directly, with seaborn we
instead give it a dataframe with the <code>data</code> argument, then
tell seaborn which columns of the dataframe we want to be used for each
axis.</p>
<p>Notice that we don’t have to give reshaped data to seaborn, it
aggregates our data for us. However, we still have to do more
complicated transformations ourselves.</p>
</section><section id="scenario-log-foldchange-scatterplot"><h2 class="section-heading">Scenario: log foldchange scatterplot<a class="anchor" aria-label="anchor" href="#scenario-log-foldchange-scatterplot"></a>
</h2>
<hr class="half-width"><p>Imagine we want to compare how different genes behave at different
timepoints, and we are expecially interested if genes on any chromosome
or of a specific type stnad out. Let’s make a scatterplot of the
log-foldchange of our genes step-by-step.</p>
<div class="section level3">
<h3 id="prepare-the-data">Prepare the data<a class="anchor" aria-label="anchor" href="#prepare-the-data"></a></h3>
<p>First, we need to aggregate our data. We could join
<code>expression_matrix</code> with the columns we want from
<code>rnaseq_df</code>, but instead let’s re-pivot our data and bring
<code>gene_biotype</code> and <code>chromosome_name</code> along.</p>
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">PYTHON<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode python" tabindex="0"><code class="sourceCode python"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>time_matrix <span class="op">=</span> rnaseq_df.pivot_table(columns <span class="op">=</span> <span class="st">"time"</span>, values <span class="op">=</span> <span class="st">"log_exp"</span>, index<span class="op">=</span>[<span class="st">"gene"</span>, <span class="st">"chromosome_name"</span>, <span class="st">"gene_biotype"</span>])</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>time_matrix <span class="op">=</span> time_matrix.reset_index(level<span class="op">=</span>[<span class="st">"chromosome_name"</span>, <span class="st">"gene_biotype"</span>])</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(time_matrix)</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>time     chromosome_name    gene_biotype          0          4          8
gene                                                                     
AI504432               3          lncRNA  10.010544  10.085962   9.974814
AW046200               8          lncRNA   7.275920   7.226188   6.346325
AW551984               9  protein_coding   7.860330   7.264752   8.194396
Aamp                   1  protein_coding  12.159714  12.235069  12.198784
Abca12                 1  protein_coding   2.509810   2.336441   2.329376
...                  ...             ...        ...        ...        ...
Zkscan3               13  protein_coding  11.031923  10.946440  10.696602
Zranb1                 7  protein_coding  12.546820  12.646836  12.984985
Zranb3                 1  protein_coding   7.582472   7.611006   7.578124
Zscan22                7  protein_coding   9.239555   9.049392   8.689101
Zw10                   9  protein_coding  10.586751  10.598198  10.524456

[1474 rows x 5 columns]</code></pre>
</div>
<div id="dataframes-practice" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="dataframes-practice" class="callout-inner">
<h3 class="callout-title">Dataframes practice<a class="anchor" aria-label="anchor" href="#dataframes-practice"></a>
</h3>
<div class="callout-content">
<p>Starting with <code>time_matrix</code>, add new columns for the log
foldchange from 0 to 4 hours and one with the log foldchange from 0 to 8
hours.</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" data-bs-parent="#accordionSolution1" aria-labelledby="headingSolution1">
<div class="accordion-body">
<p>We already have the mean log expression each gene at each timepoint.
We need to calculate the foldchange.</p>
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">PYTHON<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode python" tabindex="0"><code class="sourceCode python"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="co"># We have to use loc here since our column name is a number</span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>time_matrix[<span class="st">"logfc_0_4"</span>] <span class="op">=</span> time_matrix.loc[:,<span class="dv">4</span>] <span class="op">-</span> time_matrix.loc[:,<span class="dv">0</span>]</span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>time_matrix[<span class="st">"logfc_0_8"</span>] <span class="op">=</span> time_matrix.loc[:,<span class="dv">8</span>] <span class="op">-</span> time_matrix.loc[:,<span class="dv">0</span>]</span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(time_matrix)</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>time     chromosome_name    gene_biotype          0          4          8  \
gene                                                                        
AI504432               3          lncRNA  10.010544  10.085962   9.974814   
AW046200               8          lncRNA   7.275920   7.226188   6.346325   
AW551984               9  protein_coding   7.860330   7.264752   8.194396   
Aamp                   1  protein_coding  12.159714  12.235069  12.198784   
Abca12                 1  protein_coding   2.509810   2.336441   2.329376   
...                  ...             ...        ...        ...        ...   
Zkscan3               13  protein_coding  11.031923  10.946440  10.696602   
Zranb1                 7  protein_coding  12.546820  12.646836  12.984985   
Zranb3                 1  protein_coding   7.582472   7.611006   7.578124   
Zscan22                7  protein_coding   9.239555   9.049392   8.689101   
Zw10                   9  protein_coding  10.586751  10.598198  10.524456   

time      logfc_0_4  logfc_0_8  
gene                            
AI504432   0.075419  -0.035730  
AW046200  -0.049732  -0.929596  
AW551984  -0.595578   0.334066  
Aamp       0.075354   0.039070  
Abca12    -0.173369  -0.180433  
...             ...        ...  
Zkscan3   -0.085483  -0.335321  
Zranb1     0.100015   0.438165  
Zranb3     0.028535  -0.004348  
Zscan22   -0.190164  -0.550454  
Zw10       0.011447  -0.062295  

[1474 rows x 7 columns]</code></pre>
</div>
<p>Note that we could have first calculate foldchange and then log
transform it, but performing subtraction is more precise than division
in most programming languages, so log transforming first is
preferred.</p>
</div>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="making-an-effective-scatterplot">Making an effective scatterplot<a class="anchor" aria-label="anchor" href="#making-an-effective-scatterplot"></a></h3>
<p>Now we can make a scatterplot of the foldchanges.</p>
<div class="codewrapper sourceCode" id="cb16">
<h3 class="code-label">PYTHON<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode python" tabindex="0"><code class="sourceCode python"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>sns.scatterplot(data <span class="op">=</span> time_matrix, x <span class="op">=</span> <span class="st">"logfc_0_4"</span>, y <span class="op">=</span> <span class="st">"logfc_0_8"</span>)</span></code></pre>
</div>
<figure><img src="fig/13_8_scatter1.png" alt="" class="figure mx-auto d-block"><figcaption>Basic foldchange scatterplot.</figcaption></figure><p>Let’s now improve this plot step-by-step. First we can give the axes
more human-readable names than the dataframe’s column names.</p>
<div class="codewrapper sourceCode" id="cb17">
<h3 class="code-label">PYTHON<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode python" tabindex="0"><code class="sourceCode python"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>sns.scatterplot(data <span class="op">=</span> time_matrix, x <span class="op">=</span> <span class="st">"logfc_0_4"</span>, y <span class="op">=</span> <span class="st">"logfc_0_8"</span>)</span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>plt.xlabel(<span class="st">"0min to 4min (log FC)"</span>)</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>plt.ylabel(<span class="st">"0min to 8min (log FC)"</span>)</span></code></pre>
</div>
<figure><img src="fig/13_9_scatter2.png" alt="" class="figure mx-auto d-block"><figcaption>Axis labels changed</figcaption></figure><p>It is difficult to see see the density of genes due to there being so
many points. Lets make the points a little transparent to help with this
by changing their <code>alpha</code> level.</p>
<div class="codewrapper sourceCode" id="cb18">
<h3 class="code-label">PYTHON<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode python" tabindex="0"><code class="sourceCode python"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>sns.scatterplot(data <span class="op">=</span> time_matrix, x <span class="op">=</span> <span class="st">"logfc_0_4"</span>, y <span class="op">=</span> <span class="st">"logfc_0_8"</span>, alpha <span class="op">=</span> <span class="fl">0.4</span>)</span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>plt.xlabel(<span class="st">"0min to 4min (log FC)"</span>)</span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>plt.ylabel(<span class="st">"0min to 8min (log FC)"</span>)</span></code></pre>
</div>
<figure><img src="fig/13_10_scatter3.png" alt="" class="figure mx-auto d-block"><figcaption>Adjusted alpha level</figcaption></figure><p>Or we could make the points smaller by adjusting the <code>s</code>
argument.</p>
<div class="codewrapper sourceCode" id="cb19">
<h3 class="code-label">PYTHON<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode python" tabindex="0"><code class="sourceCode python"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>sns.scatterplot(data <span class="op">=</span> time_matrix, x <span class="op">=</span> <span class="st">"logfc_0_4"</span>, y <span class="op">=</span> <span class="st">"logfc_0_8"</span>, alpha <span class="op">=</span> <span class="fl">0.6</span>, s<span class="op">=</span><span class="dv">4</span>)</span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>plt.xlabel(<span class="st">"0min to 4min (log FC)"</span>)</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>plt.ylabel(<span class="st">"0min to 8min (log FC)"</span>)</span></code></pre>
</div>
<figure><img src="fig/13_11_scatter4.png" alt="" class="figure mx-auto d-block"><figcaption>Adjusted point size</figcaption></figure><p>Now let’s incorporate <code>gene_biotype</code>. To do this, we can
set the <code>hue</code> of the scatterplot to the column name. We’re
also going to make some more room for the plot by changing matplotlib’s
<code>rcParams</code>, which are the global plotting settings.</p>
<div id="global-plot-settings-with-rcparams" class="callout callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="global-plot-settings-with-rcparams" class="callout-inner">
<h3 class="callout-title">Global plot settings with
<code>rcParams</code><a class="anchor" aria-label="anchor" href="#global-plot-settings-with-rcparams"></a>
</h3>
<div class="callout-content">
<p><code>rcParams</code> is a specialized dictionary used by
<code>matplotlib</code> to store all global plot settings. You can
change <code>rcParams</code> to set things like the</p>
</div>
</div>
</div>
<div class="codewrapper sourceCode" id="cb20">
<h3 class="code-label">PYTHON<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode python" tabindex="0"><code class="sourceCode python"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>plt.rcParams[<span class="st">'figure.figsize'</span>] <span class="op">=</span> [<span class="dv">12</span>, <span class="dv">8</span>]</span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>sns.scatterplot(data <span class="op">=</span> time_matrix, x <span class="op">=</span> <span class="st">"logfc_0_4"</span>, y <span class="op">=</span> <span class="st">"logfc_0_8"</span>, alpha <span class="op">=</span> <span class="fl">0.6</span>, hue<span class="op">=</span><span class="st">"gene_biotype"</span>)</span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>plt.xlabel(<span class="st">"0min to 4min (log FC)"</span>)</span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a>plt.ylabel(<span class="st">"0min to 8min (log FC)"</span>)</span></code></pre>
</div>
<figure><img src="fig/13_12_scatter5.png" alt="" class="figure mx-auto d-block"><figcaption>Added gene_biotype as hue</figcaption></figure><p>It looks like we overwhelmingly have protein coding genes in the
dataset. Instead of plotting every biotype, let’s just plot whether or
not the genes are protein coding.</p>
<div id="challenge2" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Challenge<a class="anchor" aria-label="anchor" href="#challenge2"></a>
</h3>
<div class="callout-content">
<p>Create a new boolean column, <code>is_protein_coding</code> which
specifies whether or not each gene is protein coding.</p>
<p>Make a scatterplot where the <code>style</code> of the points varies
with <code>is_protein_coding</code>.</p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" data-bs-parent="#accordionSolution2" aria-labelledby="headingSolution2">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb21">
<h3 class="code-label">PYTHON<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode python" tabindex="0"><code class="sourceCode python"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>time_matrix[<span class="st">"is_protein_coding"</span>] <span class="op">=</span> time_matrix[<span class="st">"gene_biotype"</span>]<span class="op">==</span><span class="st">"protein_coding"</span></span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(time_matrix)</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>time     chromosome_name    gene_biotype          0          4          8  \
gene                                                                        
AI504432               3          lncRNA  10.010544  10.085962   9.974814   
AW046200               8          lncRNA   7.275920   7.226188   6.346325   
AW551984               9  protein_coding   7.860330   7.264752   8.194396   
Aamp                   1  protein_coding  12.159714  12.235069  12.198784   
Abca12                 1  protein_coding   2.509810   2.336441   2.329376   
...                  ...             ...        ...        ...        ...   
Zkscan3               13  protein_coding  11.031923  10.946440  10.696602   
Zranb1                 7  protein_coding  12.546820  12.646836  12.984985   
Zranb3                 1  protein_coding   7.582472   7.611006   7.578124   
Zscan22                7  protein_coding   9.239555   9.049392   8.689101   
Zw10                   9  protein_coding  10.586751  10.598198  10.524456   

time      logfc_0_4  logfc_0_8  is_protein_coding  
gene                                               
AI504432   0.075419  -0.035730              False  
AW046200  -0.049732  -0.929596              False  
AW551984  -0.595578   0.334066               True  
Aamp       0.075354   0.039070               True  
Abca12    -0.173369  -0.180433               True  
...             ...        ...                ...  
Zkscan3   -0.085483  -0.335321               True  
Zranb1     0.100015   0.438165               True  
Zranb3     0.028535  -0.004348               True  
Zscan22   -0.190164  -0.550454               True  
Zw10       0.011447  -0.062295               True  

[1474 rows x 8 columns]</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb23">
<h3 class="code-label">PYTHON<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode python" tabindex="0"><code class="sourceCode python"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>sns.scatterplot(data <span class="op">=</span> time_matrix, x <span class="op">=</span> <span class="st">"logfc_0_4"</span>, y <span class="op">=</span> <span class="st">"logfc_0_8"</span>, alpha <span class="op">=</span> <span class="fl">0.6</span>, style<span class="op">=</span><span class="st">"is_protein_coding"</span>)</span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>plt.xlabel(<span class="st">"0min to 4min (log FC)"</span>)</span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a>plt.ylabel(<span class="st">"0min to 8min (log FC)"</span>)</span></code></pre>
</div>
<figure><img src="fig/13_13_scatter6.png" alt="" class="figure mx-auto d-block"><figcaption>Style set to is_protein_coding</figcaption></figure></div>
</div>
</div>
</div>
<p>Now that hue is freed up, we can use it to encode for chromosome.</p>
<div class="codewrapper sourceCode" id="cb24">
<h3 class="code-label">PYTHON<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode python" tabindex="0"><code class="sourceCode python"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="co">#This hit my personal stylistic limit, so I'm spread the call to multiple lines</span></span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>sns.scatterplot(data <span class="op">=</span> time_matrix, </span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a>                x <span class="op">=</span> <span class="st">"logfc_0_4"</span>, </span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a>                y <span class="op">=</span> <span class="st">"logfc_0_8"</span>, </span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a>                alpha <span class="op">=</span> <span class="fl">0.6</span>, </span>
<span id="cb24-6"><a href="#cb24-6" aria-hidden="true" tabindex="-1"></a>                style <span class="op">=</span> <span class="st">"is_protein_coding"</span>,</span>
<span id="cb24-7"><a href="#cb24-7" aria-hidden="true" tabindex="-1"></a>                hue <span class="op">=</span> <span class="st">"chromosome_name"</span>)</span>
<span id="cb24-8"><a href="#cb24-8" aria-hidden="true" tabindex="-1"></a>plt.xlabel(<span class="st">"0min to 4min (log FC)"</span>)</span>
<span id="cb24-9"><a href="#cb24-9" aria-hidden="true" tabindex="-1"></a>plt.ylabel(<span class="st">"0min to 8min (log FC)"</span>)</span></code></pre>
</div>
<figure><img src="fig/13_14_scatter7.png" alt="" class="figure mx-auto d-block"><figcaption>Adding chromosome</figcaption></figure><p>There doesn’t seem to be any interesting pattern. Let’s go back to
just plotting whether or not the gene is protein coding, and do a little
more to clean up the plot.</p>
<div class="codewrapper sourceCode" id="cb25">
<h3 class="code-label">PYTHON<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode python" tabindex="0"><code class="sourceCode python"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="co"># This scales all text in the plot by 150%</span></span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>sns.<span class="bu">set</span>(font_scale <span class="op">=</span> <span class="fl">1.5</span>)</span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a>sns.scatterplot(data <span class="op">=</span> time_matrix, </span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a>                x <span class="op">=</span> <span class="st">"logfc_0_4"</span>, </span>
<span id="cb25-6"><a href="#cb25-6" aria-hidden="true" tabindex="-1"></a>                y <span class="op">=</span> <span class="st">"logfc_0_8"</span>, </span>
<span id="cb25-7"><a href="#cb25-7" aria-hidden="true" tabindex="-1"></a>                alpha <span class="op">=</span> <span class="fl">0.6</span>, </span>
<span id="cb25-8"><a href="#cb25-8" aria-hidden="true" tabindex="-1"></a>                hue <span class="op">=</span> <span class="st">"is_protein_coding"</span>, </span>
<span id="cb25-9"><a href="#cb25-9" aria-hidden="true" tabindex="-1"></a>                <span class="co">#Sets the color pallete to map values to</span></span>
<span id="cb25-10"><a href="#cb25-10" aria-hidden="true" tabindex="-1"></a>                palette <span class="op">=</span> <span class="st">"mako"</span>)</span>
<span id="cb25-11"><a href="#cb25-11" aria-hidden="true" tabindex="-1"></a>plt.xlabel(<span class="st">"0min to 4min (log FC)"</span>)</span>
<span id="cb25-12"><a href="#cb25-12" aria-hidden="true" tabindex="-1"></a>plt.ylabel(<span class="st">"0min to 8min (log FC)"</span>)</span>
<span id="cb25-13"><a href="#cb25-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-14"><a href="#cb25-14" aria-hidden="true" tabindex="-1"></a><span class="co"># Set a legend title and give it 2 columns</span></span>
<span id="cb25-15"><a href="#cb25-15" aria-hidden="true" tabindex="-1"></a>plt.legend(ncol <span class="op">=</span> <span class="dv">2</span>, title<span class="op">=</span><span class="st">"Is Protein Coding"</span>)</span></code></pre>
</div>
<figure><img src="fig/13_15_scatter.png" alt="" class="figure mx-auto d-block"><figcaption>Cleaning up the plot</figcaption></figure></div>
</section><section id="heatmaps"><h2 class="section-heading">Heatmaps<a class="anchor" aria-label="anchor" href="#heatmaps"></a>
</h2>
<hr class="half-width"><p>Seaborn also has great built-in support for heatmaps.</p>
<p>We can make heatmaps using <code>heatmap</code> or
<code>clustermap</code>, depending on if we want our data to be
clustered. Let’s take a look at the correlation matrix for our
samples.</p>
<div class="codewrapper sourceCode" id="cb26">
<h3 class="code-label">PYTHON<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode python" tabindex="0"><code class="sourceCode python"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>sample_matrix <span class="op">=</span> rnaseq_df.pivot_table(columns <span class="op">=</span> <span class="st">"sample"</span>, values <span class="op">=</span> <span class="st">"log_exp"</span>, index<span class="op">=</span><span class="st">"gene"</span>)</span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(sample_matrix.corr())</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>sample      GSM2545336  GSM2545337  GSM2545338  GSM2545339  GSM2545340  \
sample                                                                   
GSM2545336    1.000000    0.972832    0.970700    0.979381    0.957032   
GSM2545337    0.972832    1.000000    0.992230    0.983430    0.972068   
GSM2545338    0.970700    0.992230    1.000000    0.981653    0.970098   
GSM2545339    0.979381    0.983430    0.981653    1.000000    0.961275   
GSM2545340    0.957032    0.972068    0.970098    0.961275    1.000000   
GSM2545341    0.957018    0.951554    0.950648    0.940430    0.979411   
GSM2545342    0.987712    0.980026    0.978246    0.974903    0.968918   
GSM2545343    0.948861    0.971826    0.973301    0.959663    0.988114   
GSM2545344    0.973664    0.987626    0.985568    0.980757    0.970084   
GSM2545345    0.943748    0.963804    0.963715    0.946226    0.984994   
GSM2545346    0.967839    0.961595    0.957958    0.955665    0.984706   
GSM2545347    0.958351    0.957363    0.955393    0.944399    0.982981   
GSM2545348    0.973558    0.992018    0.991673    0.982984    0.971515   
GSM2545349    0.948938    0.970211    0.971276    0.955778    0.988263   
GSM2545350    0.952803    0.973764    0.971737    0.960796    0.991229   
GSM2545351    0.991921    0.977846    0.973363    0.979936    0.962843   
GSM2545352    0.980437    0.989274    0.986909    0.983319    0.974796   
GSM2545353    0.974628    0.989940    0.989857    0.982648    0.971325   
GSM2545354    0.950621    0.972800    0.972051    0.959123    0.987231   
GSM2545362    0.986494    0.981655    0.978767    0.988969    0.962088   
GSM2545363    0.936577    0.962115    0.962892    0.940839    0.980774   
GSM2545380    0.988869    0.977962    0.975382    0.973683    0.965520   

sample      GSM2545341  GSM2545342  GSM2545343  GSM2545344  GSM2545345  ...  \
sample                                                                  ...   
GSM2545336    0.957018    0.987712    0.948861    0.973664    0.943748  ...   
GSM2545337    0.951554    0.980026    0.971826    0.987626    0.963804  ...   
GSM2545338    0.950648    0.978246    0.973301    0.985568    0.963715  ...   
GSM2545339    0.940430    0.974903    0.959663    0.980757    0.946226  ...   
GSM2545340    0.979411    0.968918    0.988114    0.970084    0.984994  ...   
GSM2545341    1.000000    0.967231    0.967720    0.958880    0.978091  ...   
GSM2545342    0.967231    1.000000    0.959923    0.980462    0.957994  ...   
GSM2545343    0.967720    0.959923    1.000000    0.965428    0.983462  ...   
GSM2545344    0.958880    0.980462    0.965428    1.000000    0.967708  ...   
GSM2545345    0.978091    0.957994    0.983462    0.967708    1.000000  ...   
GSM2545346    0.983828    0.973720    0.977478    0.962581    0.976912  ...   
GSM2545347    0.986949    0.968972    0.974794    0.962113    0.981518  ...   
GSM2545348    0.950961    0.980532    0.974156    0.986710    0.965095  ...   
GSM2545349    0.971058    0.960674    0.992085    0.966662    0.986116  ...   
GSM2545350    0.973947    0.964523    0.990495    0.970990    0.986146  ...   
GSM2545351    0.961638    0.989280    0.953650    0.979976    0.950451  ...   
GSM2545352    0.960888    0.985541    0.971954    0.990066    0.968691  ...   
GSM2545353    0.952424    0.982904    0.971896    0.989196    0.965436  ...   
GSM2545354    0.973524    0.961077    0.988615    0.972600    0.987209  ...   
GSM2545362    0.951186    0.982616    0.955145    0.983885    0.946956  ...   
GSM2545363    0.970944    0.952557    0.980218    0.965252    0.987553  ...   
GSM2545380    0.963521    0.990511    0.957152    0.978371    0.956813  ...   

sample      GSM2545348  GSM2545349  GSM2545350  GSM2545351  GSM2545352  \
sample                                                                   
GSM2545336    0.973558    0.948938    0.952803    0.991921    0.980437   
GSM2545337    0.992018    0.970211    0.973764    0.977846    0.989274   
GSM2545338    0.991673    0.971276    0.971737    0.973363    0.986909   
GSM2545339    0.982984    0.955778    0.960796    0.979936    0.983319   
GSM2545340    0.971515    0.988263    0.991229    0.962843    0.974796   
GSM2545341    0.950961    0.971058    0.973947    0.961638    0.960888   
GSM2545342    0.980532    0.960674    0.964523    0.989280    0.985541   
GSM2545343    0.974156    0.992085    0.990495    0.953650    0.971954   
GSM2545344    0.986710    0.966662    0.970990    0.979976    0.990066   
GSM2545345    0.965095    0.986116    0.986146    0.950451    0.968691   
GSM2545346    0.960875    0.978673    0.981337    0.971079    0.970621   
GSM2545347    0.958524    0.978974    0.980361    0.964883    0.967565   
GSM2545348    1.000000    0.971489    0.973997    0.977225    0.990905   
GSM2545349    0.971489    1.000000    0.990858    0.954456    0.971922   
GSM2545350    0.973997    0.990858    1.000000    0.959045    0.975085   
GSM2545351    0.977225    0.954456    0.959045    1.000000    0.983727   
GSM2545352    0.990905    0.971922    0.975085    0.983727    1.000000   
GSM2545353    0.992297    0.969924    0.972310    0.979291    0.991556   
GSM2545354    0.972151    0.990242    0.989228    0.956395    0.972436   
GSM2545362    0.980224    0.955712    0.959541    0.988591    0.984683   
GSM2545363    0.963226    0.983872    0.982521    0.943499    0.966163   
GSM2545380    0.979398    0.956662    0.961779    0.989683    0.985750   

sample      GSM2545353  GSM2545354  GSM2545362  GSM2545363  GSM2545380  
sample                                                                  
GSM2545336    0.974628    0.950621    0.986494    0.936577    0.988869  
GSM2545337    0.989940    0.972800    0.981655    0.962115    0.977962  
GSM2545338    0.989857    0.972051    0.978767    0.962892    0.975382  
GSM2545339    0.982648    0.959123    0.988969    0.940839    0.973683  
GSM2545340    0.971325    0.987231    0.962088    0.980774    0.965520  
GSM2545341    0.952424    0.973524    0.951186    0.970944    0.963521  
GSM2545342    0.982904    0.961077    0.982616    0.952557    0.990511  
GSM2545343    0.971896    0.988615    0.955145    0.980218    0.957152  
GSM2545344    0.989196    0.972600    0.983885    0.965252    0.978371  
GSM2545345    0.965436    0.987209    0.946956    0.987553    0.956813  
GSM2545346    0.963956    0.975386    0.961966    0.967480    0.973104  
GSM2545347    0.961118    0.977208    0.952988    0.972364    0.967319  
GSM2545348    0.992297    0.972151    0.980224    0.963226    0.979398  
GSM2545349    0.969924    0.990242    0.955712    0.983872    0.956662  
GSM2545350    0.972310    0.989228    0.959541    0.982521    0.961779  
GSM2545351    0.979291    0.956395    0.988591    0.943499    0.989683  
GSM2545352    0.991556    0.972436    0.984683    0.966163    0.985750  
GSM2545353    1.000000    0.971512    0.980850    0.962497    0.981468  
GSM2545354    0.971512    1.000000    0.958851    0.983429    0.957011  
GSM2545362    0.980850    0.958851    1.000000    0.942461    0.980664  
GSM2545363    0.962497    0.983429    0.942461    1.000000    0.951069  
GSM2545380    0.981468    0.957011    0.980664    0.951069    1.000000  

[22 rows x 22 columns]</code></pre>
</div>
<p>We can see the structure of these correlations in a clustered
heatmap.</p>
<div class="codewrapper sourceCode" id="cb28">
<h3 class="code-label">PYTHON<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode python" tabindex="0"><code class="sourceCode python"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>sns.clustermap(sample_matrix.corr())</span></code></pre>
</div>
<figure><img src="fig/13_16_heatmap1.png" alt="" class="figure mx-auto d-block"><figcaption>Default clustered heatmap</figcaption></figure><p>We can change the color mapping.</p>
<div class="codewrapper sourceCode" id="cb29">
<h3 class="code-label">PYTHON<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode python" tabindex="0"><code class="sourceCode python"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>sns.clustermap(sample_matrix.corr(), cmap <span class="op">=</span> <span class="st">"magma"</span>)</span></code></pre>
</div>
<figure><img src="fig/13_17_heatmap2.png" alt="" class="figure mx-auto d-block"><figcaption>Using magma colormap</figcaption></figure><p>We can add sample information to the plot by changing the index.
First we need to get the data.</p>
<div class="codewrapper sourceCode" id="cb30">
<h3 class="code-label">PYTHON<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode python" tabindex="0"><code class="sourceCode python"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="co"># We select all of the columns related to the samples we might want to look at, group by sample, then aggregate by the most common value, or mode</span></span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a>sample_data <span class="op">=</span> rnaseq_df[[<span class="st">'sample'</span>,<span class="st">'organism'</span>, <span class="st">'age'</span>, <span class="st">'sex'</span>, <span class="st">'infection'</span>,</span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a>       <span class="st">'strain'</span>, <span class="st">'time'</span>, <span class="st">'tissue'</span>, <span class="st">'mouse'</span>]].groupby(<span class="st">'sample'</span>).agg(pd.Series.mode)</span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(sample_data)</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>	organism 	age 	sex 	infection 	strain 	time 	tissue 	mouse
sample 								
GSM2545336 	Mus musculus 	8 	Female 	InfluenzaA 	C57BL/6 	8 	Cerebellum 	14
GSM2545337 	Mus musculus 	8 	Female 	NonInfected 	C57BL/6 	0 	Cerebellum 	9
GSM2545338 	Mus musculus 	8 	Female 	NonInfected 	C57BL/6 	0 	Cerebellum 	10
GSM2545339 	Mus musculus 	8 	Female 	InfluenzaA 	C57BL/6 	4 	Cerebellum 	15
GSM2545340 	Mus musculus 	8 	Male 	InfluenzaA 	C57BL/6 	4 	Cerebellum 	18
GSM2545341 	Mus musculus 	8 	Male 	InfluenzaA 	C57BL/6 	8 	Cerebellum 	6
GSM2545342 	Mus musculus 	8 	Female 	InfluenzaA 	C57BL/6 	8 	Cerebellum 	5
GSM2545343 	Mus musculus 	8 	Male 	NonInfected 	C57BL/6 	0 	Cerebellum 	11
GSM2545344 	Mus musculus 	8 	Female 	InfluenzaA 	C57BL/6 	4 	Cerebellum 	22
GSM2545345 	Mus musculus 	8 	Male 	InfluenzaA 	C57BL/6 	4 	Cerebellum 	13
GSM2545346 	Mus musculus 	8 	Male 	InfluenzaA 	C57BL/6 	8 	Cerebellum 	23
GSM2545347 	Mus musculus 	8 	Male 	InfluenzaA 	C57BL/6 	8 	Cerebellum 	24
GSM2545348 	Mus musculus 	8 	Female 	NonInfected 	C57BL/6 	0 	Cerebellum 	8
GSM2545349 	Mus musculus 	8 	Male 	NonInfected 	C57BL/6 	0 	Cerebellum 	7
GSM2545350 	Mus musculus 	8 	Male 	InfluenzaA 	C57BL/6 	4 	Cerebellum 	1
GSM2545351 	Mus musculus 	8 	Female 	InfluenzaA 	C57BL/6 	8 	Cerebellum 	16
GSM2545352 	Mus musculus 	8 	Female 	InfluenzaA 	C57BL/6 	4 	Cerebellum 	21
GSM2545353 	Mus musculus 	8 	Female 	NonInfected 	C57BL/6 	0 	Cerebellum 	4
GSM2545354 	Mus musculus 	8 	Male 	NonInfected 	C57BL/6 	0 	Cerebellum 	2
GSM2545362 	Mus musculus 	8 	Female 	InfluenzaA 	C57BL/6 	4 	Cerebellum 	20
GSM2545363 	Mus musculus 	8 	Male 	InfluenzaA 	C57BL/6 	4 	Cerebellum 	12
GSM2545380 	Mus musculus 	8 	Female 	InfluenzaA 	C57BL/6 	8 	Cerebellum 	19</code></pre>
</div>
<p>Then we can join, set the index, and plot to see the sample
groupings.</p>
<div class="codewrapper sourceCode" id="cb32">
<h3 class="code-label">PYTHON<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode python" tabindex="0"><code class="sourceCode python"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a>corr_df <span class="op">=</span> sample_matrix.corr()</span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a>corr_df <span class="op">=</span> corr_df.join(sample_data)</span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a>corr_df <span class="op">=</span> corr_df.set_index(<span class="st">"infection"</span>)</span>
<span id="cb32-4"><a href="#cb32-4" aria-hidden="true" tabindex="-1"></a>sns.clustermap(corr_df.iloc[:,:<span class="op">-</span><span class="dv">8</span>],</span>
<span id="cb32-5"><a href="#cb32-5" aria-hidden="true" tabindex="-1"></a>               cmap <span class="op">=</span> <span class="st">"magma"</span>)</span></code></pre>
</div>
<figure><img src="fig/13_18_heatmap3.png" alt="" class="figure mx-auto d-block"><figcaption>Heatmap with infection information</figcaption></figure><div id="creating-a-boxplot" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="creating-a-boxplot" class="callout-inner">
<h3 class="callout-title">Creating a boxplot<a class="anchor" aria-label="anchor" href="#creating-a-boxplot"></a>
</h3>
<div class="callout-content">
<p>Visualize log mean gene expression by sample and some other variable
of choice in <code>rnaseq_df</code>. Take some time to try to get this
plot close to publication ready for either a poster, presentation, or
paper.</p>
</div>
</div>
</div>
<div id="accordionHint1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button hint-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseHint1" aria-expanded="false" aria-controls="collapseHint1">
  <h4 class="accordion-header" id="headingHint1">
  Give me a hint
  </h4>
</button>
<div id="collapseHint1" class="accordion-collapse collapse" data-bs-parent="#accordionHint1" aria-labelledby="headingHint1">
<div class="accordion-body">
<p>You may need want to rotate the x labels of the plot, which can be
done with <code>plt.xticks(rotation = degrees)</code> where
<code>degrees</code> is the number of degrees you wish to rotate by.</p>
</div>
</div>
</div>
</div>
<div id="accordionHint2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button hint-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseHint2" aria-expanded="false" aria-controls="collapseHint2">
  <h4 class="accordion-header" id="headingHint2">
  Give me a hint
  </h4>
</button>
<div id="collapseHint2" class="accordion-collapse collapse" data-bs-parent="#accordionHint2" aria-labelledby="headingHint2">
<div class="accordion-body">
<p>Check out some of the details of <code>sns.set_context()</code> <a href="https://seaborn.pydata.org/tutorial/aesthetics.html#scaling-plot-elements" class="external-link">here</a>.
Seaborn has useful size default for different figure types.</p>
</div>
</div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" data-bs-parent="#accordionSolution3" aria-labelledby="headingSolution3">
<div class="accordion-body">
<p>Your plot will of course vary, but here are some examples of what we
can do:</p>
<div class="codewrapper sourceCode" id="cb33">
<h3 class="code-label">PYTHON<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode python" tabindex="0"><code class="sourceCode python"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a>sns.set_context(<span class="st">"talk"</span>)</span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Things like this will depend on your screen resolution and size</span></span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a><span class="co">#plt.rcParams['figure.figsize'] = [8, 6]</span></span>
<span id="cb33-4"><a href="#cb33-4" aria-hidden="true" tabindex="-1"></a><span class="co">#plt.rcParams['figure.dpi'] = 300</span></span>
<span id="cb33-5"><a href="#cb33-5" aria-hidden="true" tabindex="-1"></a><span class="co">#sns.set(font_scale=1.25)</span></span>
<span id="cb33-6"><a href="#cb33-6" aria-hidden="true" tabindex="-1"></a>sns.set_palette(sns.color_palette(<span class="st">"rocket_r"</span>, n_colors<span class="op">=</span><span class="dv">3</span>))</span>
<span id="cb33-7"><a href="#cb33-7" aria-hidden="true" tabindex="-1"></a>sns.boxplot(rnaseq_df, </span>
<span id="cb33-8"><a href="#cb33-8" aria-hidden="true" tabindex="-1"></a>            x <span class="op">=</span> <span class="st">"sample"</span>, </span>
<span id="cb33-9"><a href="#cb33-9" aria-hidden="true" tabindex="-1"></a>            y <span class="op">=</span> <span class="st">"log_exp"</span>,</span>
<span id="cb33-10"><a href="#cb33-10" aria-hidden="true" tabindex="-1"></a>            hue <span class="op">=</span> <span class="st">"time"</span>, </span>
<span id="cb33-11"><a href="#cb33-11" aria-hidden="true" tabindex="-1"></a>            dodge<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb33-12"><a href="#cb33-12" aria-hidden="true" tabindex="-1"></a>plt.xticks(rotation<span class="op">=</span><span class="dv">45</span>, ha<span class="op">=</span><span class="st">'right'</span>)<span class="op">;</span></span>
<span id="cb33-13"><a href="#cb33-13" aria-hidden="true" tabindex="-1"></a>plt.ylabel(<span class="st">"Log Expression"</span>)</span>
<span id="cb33-14"><a href="#cb33-14" aria-hidden="true" tabindex="-1"></a>plt.xlabel(<span class="st">""</span>)</span>
<span id="cb33-15"><a href="#cb33-15" aria-hidden="true" tabindex="-1"></a><span class="co"># See this thread for more discussion of legend positions: https://stackoverflow.com/a/43439132</span></span>
<span id="cb33-16"><a href="#cb33-16" aria-hidden="true" tabindex="-1"></a>plt.legend(title<span class="op">=</span><span class="st">'Days Post Infection'</span>, loc<span class="op">=</span>(<span class="fl">1.04</span>, <span class="fl">0.5</span>), labels<span class="op">=</span>[<span class="st">'0 days'</span>,<span class="st">'4 days'</span>,<span class="st">'8 days'</span>])</span></code></pre>
</div>
<figure><img src="fig/13_16_challenge1.png" class="figure mx-auto d-block"></figure><div class="codewrapper sourceCode" id="cb34">
<h3 class="code-label">PYTHON<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode python" tabindex="0"><code class="sourceCode python"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a>sns.set_palette([<span class="st">"#A51C30"</span>,<span class="st">"#8996A0"</span>])</span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a>sns.violinplot(rnaseq_df, </span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a>            x <span class="op">=</span> <span class="st">"sample"</span>, </span>
<span id="cb34-4"><a href="#cb34-4" aria-hidden="true" tabindex="-1"></a>            y <span class="op">=</span> <span class="st">"log_exp"</span>,</span>
<span id="cb34-5"><a href="#cb34-5" aria-hidden="true" tabindex="-1"></a>            hue <span class="op">=</span> <span class="st">"infection"</span>, </span>
<span id="cb34-6"><a href="#cb34-6" aria-hidden="true" tabindex="-1"></a>            dodge<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb34-7"><a href="#cb34-7" aria-hidden="true" tabindex="-1"></a>plt.xticks(rotation<span class="op">=</span><span class="dv">45</span>, ha<span class="op">=</span><span class="st">'right'</span>)<span class="op">;</span></span>
<span id="cb34-8"><a href="#cb34-8" aria-hidden="true" tabindex="-1"></a>plt.legend(loc <span class="op">=</span> <span class="st">'lower right'</span>)</span></code></pre>
</div>
<figure><img src="fig/13_17_challenge2.png" class="figure mx-auto d-block"></figure><div class="codewrapper sourceCode" id="cb35">
<h3 class="code-label">PYTHON<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode python" tabindex="0"><code class="sourceCode python"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a>sns.set_context(<span class="st">"talk"</span>)</span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a>pal <span class="op">=</span> sns.color_palette(<span class="st">"blend:#A51C30,#FFDB6D"</span>, <span class="dv">24</span>)</span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a>g <span class="op">=</span> sns.catplot(rnaseq_df,</span>
<span id="cb35-4"><a href="#cb35-4" aria-hidden="true" tabindex="-1"></a>            kind <span class="op">=</span> <span class="st">"box"</span>,</span>
<span id="cb35-5"><a href="#cb35-5" aria-hidden="true" tabindex="-1"></a>            x <span class="op">=</span> <span class="st">"sample"</span>, </span>
<span id="cb35-6"><a href="#cb35-6" aria-hidden="true" tabindex="-1"></a>            y <span class="op">=</span> <span class="st">"log_exp"</span>,</span>
<span id="cb35-7"><a href="#cb35-7" aria-hidden="true" tabindex="-1"></a>            col <span class="op">=</span> <span class="st">"sex"</span>,</span>
<span id="cb35-8"><a href="#cb35-8" aria-hidden="true" tabindex="-1"></a>            dodge<span class="op">=</span><span class="va">False</span>, </span>
<span id="cb35-9"><a href="#cb35-9" aria-hidden="true" tabindex="-1"></a>            hue <span class="op">=</span> <span class="st">"sample"</span>,</span>
<span id="cb35-10"><a href="#cb35-10" aria-hidden="true" tabindex="-1"></a>            palette <span class="op">=</span> pal,</span>
<span id="cb35-11"><a href="#cb35-11" aria-hidden="true" tabindex="-1"></a>            sharex <span class="op">=</span> <span class="va">False</span>)</span>
<span id="cb35-12"><a href="#cb35-12" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> axes <span class="kw">in</span> g.axes.flat:</span>
<span id="cb35-13"><a href="#cb35-13" aria-hidden="true" tabindex="-1"></a>    _ <span class="op">=</span> axes.set_xticklabels(axes.get_xticklabels(), rotation<span class="op">=</span><span class="dv">45</span>, ha<span class="op">=</span><span class="st">'right'</span>)</span>
<span id="cb35-14"><a href="#cb35-14" aria-hidden="true" tabindex="-1"></a>g.set_axis_labels(<span class="st">""</span>, <span class="st">"Log Expression"</span>)</span>
<span id="cb35-15"><a href="#cb35-15" aria-hidden="true" tabindex="-1"></a>plt.subplots_adjust(wspace<span class="op">=</span><span class="fl">0.1</span>)</span></code></pre>
</div>
<figure><img src="fig/13_18_challenge3.png" class="figure mx-auto d-block"></figure></div>
</div>
</div>
</div>
<div id="saving-your-plot-to-a-file" class="callout callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="saving-your-plot-to-a-file" class="callout-inner">
<h3 class="callout-title">Saving your plot to a file<a class="anchor" aria-label="anchor" href="#saving-your-plot-to-a-file"></a>
</h3>
<div class="callout-content">
<p>If you are satisfied with the plot you see you may want to save it to
a file, perhaps to include it in a publication. There is a function in
the matplotlib.pyplot module that accomplishes this: <a href="https://matplotlib.org/api/_as_gen/matplotlib.pyplot.savefig.html" class="external-link">savefig</a>.
Calling this function, e.g. with</p>
<div class="codewrapper sourceCode" id="cb36">
<h3 class="code-label">PYTHON<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode python" tabindex="0"><code class="sourceCode python"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a>plt.savefig(<span class="st">'my_figure.png'</span>)</span></code></pre>
</div>
<p>will save the current figure to the file <code>my_figure.png</code>.
The file format will automatically be deduced from the file name
extension (other formats are pdf, ps, eps and svg).</p>
<p>Note that functions in <code>plt</code> refer to a global figure
variable and after a figure has been displayed to the screen (e.g. with
<code>plt.show</code>) matplotlib will make this variable refer to a new
empty figure. Therefore, make sure you call <code>plt.savefig</code>
before the plot is displayed to the screen, otherwise you may find a
file with an empty plot.</p>
<p>When using dataframes, data is often generated and plotted to screen
in one line. In addition to using <code>plt.savefig</code>, we can save
a reference to the current figure in a local variable (with
<code>plt.gcf</code>) and call the <code>savefig</code> class method
from that variable to save the figure to file.</p>
<div class="codewrapper sourceCode" id="cb37">
<h3 class="code-label">PYTHON<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode python" tabindex="0"><code class="sourceCode python"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a>data.plot(kind<span class="op">=</span><span class="st">'bar'</span>)</span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a>fig <span class="op">=</span> plt.gcf() <span class="co"># get current figure</span></span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a>fig.savefig(<span class="st">'my_figure.png'</span>)</span></code></pre>
</div>
<p>This supports most common image formats such as <code>png</code>,
<code>svg</code>, <code>pdf</code>, etc.</p>
</div>
</div>
</div>
<div id="making-your-plots-accessible" class="callout callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="making-your-plots-accessible" class="callout-inner">
<h3 class="callout-title">Making your plots accessible<a class="anchor" aria-label="anchor" href="#making-your-plots-accessible"></a>
</h3>
<div class="callout-content">
<p>Whenever you are generating plots to go into a paper or a
presentation, there are a few things you can do to make sure that
everyone can understand your plots. * Always make sure your text is
large enough to read. Use the <code>fontsize</code> parameter in
<code>xlabel</code>, <code>ylabel</code>, <code>title</code>, and
<code>legend</code>, and <a href="https://matplotlib.org/2.1.1/api/_as_gen/matplotlib.pyplot.tick_params.html" class="external-link"><code>tick_params</code>
with <code>labelsize</code></a> to increase the text size of the numbers
on your axes. * Similarly, you should make your graph elements easy to
see. Use <code>s</code> to increase the size of your scatterplot markers
and <code>linewidth</code> to increase the sizes of your plot lines. *
Using color (and nothing else) to distinguish between different plot
elements will make your plots unreadable to anyone who is colorblind, or
who happens to have a black-and-white office printer. For lines, the
<code>linestyle</code> parameter lets you use different types of lines.
For scatterplots, <code>marker</code> lets you change the shape of your
points. If you’re unsure about your colors, you can use <a href="https://www.color-blindness.com/coblis-color-blindness-simulator/" class="external-link">Coblis</a>
or <a href="https://colororacle.org/" class="external-link">Color Oracle</a> to simulate what
your plots would look like to those with colorblindness.</p>
</div>
</div>
</div>
<!-- 
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use. 
 -->
</section></div> <!-- / div.lesson-content -->
    </main><!-- / main#main-content.main-content --><nav class="bottom-pagination mx-md-4" aria-label="Previous and Next Chapter"><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="home-06-combining-data.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="14-sklearn.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="home-06-combining-data.html" rel="prev"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous: Combining Data</a>
        <a class="chapter-link float-end" href="14-sklearn.html" rel="next">Next: Perform machine... <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
    </nav></div> <!-- / div.primary-content.col-xs-12 -->
<!-- END:   inst/pkgdown/templates/content-instructor.html-->

      </div><!--/div.row-->
      		<footer class="row footer mx-md-3"><hr><div class="col-md-6">
				<p>This lesson is subject to the <a href="CODE_OF_CONDUCT.html">Code of Conduct</a></p>
        <p>
        
        <a href="https://github.com/ccb-hms/workbench-python-workshop/edit/main/episodes/13-plotting.md" class="external-link">Edit on GitHub</a>
        
	
        | <a href="https://github.com/ccb-hms/workbench-python-workshop/blob/main/CONTRIBUTING.md" class="external-link">Contributing</a> 
        | <a href="https://github.com/ccb-hms/workbench-python-workshop/" class="external-link">Source</a></p>
				<p><a href="https://github.com/ccb-hms/workbench-python-workshop/blob/main/CITATION" class="external-link">Cite</a> | <a href="mailto:christopher_magnano@hms.harvard.edu">Contact</a> | <a href="https://carpentries.org/about/" class="external-link">About</a></p>
			</div>
			<div class="col-md-6">
        
        <p>Materials licensed under <a href="LICENSE.html">CC-BY 4.0</a> by the authors</p>
        
        <p><a href="https://creativecommons.org/licenses/by-sa/4.0/" class="external-link">Template licensed under CC-BY 4.0</a> by <a href="https://carpentries.org" class="external-link">The Carpentries</a></p>
        <p>Built with <a href="https://github.com/carpentries/sandpaper" class="external-link">sandpaper (0.12.5)</a>,
        <a href="https://github.com/carpentries/pegboard" class="external-link">pegboard (0.5.3)</a>,
      and <a href="https://github.com/carpentries/varnish/tree/0.2.17" class="external-link">varnish (0.2.17)</a>.</p>
			</div>
		</footer></div> <!-- / div.container -->
	<div id="to-top">
		<a href="#top">
			<i class="search-icon" data-feather="arrow-up" role="img" aria-label="Back to top"></i><br><span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top
		</a>
	</div>
  <script type="application/ld+json">
    {
  "@context": "https://schema.org",
  "@type": "TrainingMaterial",
  "@id": "https://ccb-hms.github.io/workbench-python-workshop/13-plotting.html",
  "dct:conformsTo": "https://bioschemas.org/profiles/TrainingMaterial/1.0-RELEASE",
  "description": "A Carpentries Lesson teaching foundational data and coding skills to researchers worldwide",
  "keywords": "python, pandas, programming, seaborn, introductory",
  "name": "Visualizing data with matplotlib and seaborn",
  "creativeWorkStatus": "active",
  "url": "https://ccb-hms.github.io/workbench-python-workshop/13-plotting.html",
  "identifier": "https://ccb-hms.github.io/workbench-python-workshop/13-plotting.html",
  "dateCreated": "2023-04-01",
  "dateModified": "2023-04-27",
  "datePublished": "2023-07-25"
}

  </script><script>
		feather.replace();
	</script><!-- Matomo
    2022-11-07: we have gotten a notification that we have an overage for our
    tracking and I'm pretty sure this has to do with Workbench usage.
    Considering that I am not _currently_ using this tracking because I do not
    yet know how to access the data, I am turning this off for now.
  <script>
    var _paq = window._paq = window._paq || [];
    /* tracker methods like "setCustomDimension" should be called before "trackPageView" */
    _paq.push(["setDocumentTitle", document.domain + "/" + document.title]);
    _paq.push(["setDomains", ["*.preview.carpentries.org","*.datacarpentry.github.io","*.datacarpentry.org","*.librarycarpentry.github.io","*.librarycarpentry.org","*.swcarpentry.github.io", "*.carpentries.github.io"]]);
    _paq.push(["setDoNotTrack", true]);
    _paq.push(["disableCookies"]);
    _paq.push(['trackPageView']);
    _paq.push(['enableLinkTracking']);
    (function() {
          var u="https://carpentries.matomo.cloud/";
          _paq.push(['setTrackerUrl', u+'matomo.php']);
          _paq.push(['setSiteId', '1']);
          var d=document, g=d.createElement('script'), s=d.getElementsByTagName('script')[0];
          g.async=true; g.src='https://cdn.matomo.cloud/carpentries.matomo.cloud/matomo.js'; s.parentNode.insertBefore(g,s);
        })();
  </script>
  End Matomo Code --></body></html><!-- END:   inst/pkgdown/templates/layout.html-->

